compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -418,6 +438,21 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
         }
         // Recursive checking
         if let Some(ref_tracking) = self.ref_tracking.as_deref_mut() {
+            // Determine whether this pointer expects to be pointing to something mutable.
+            let ptr_expected_mutbl = match ptr_kind {
+                PointerKind::Box => Mutability::Mut,
+                PointerKind::Ref => {
+                    let tam = value.layout.ty.builtin_deref(false).unwrap();
+                    // ZST never require mutability. We do not take into account interior mutability here
+                    // since we cannot know if there really is an `UnsafeCell` -- so we check that
+                    // in the recursive descent behind this reference.
+                    if size == Size::ZERO || tam.mutbl == Mutability::Not {
+                        Mutability::Not
+                    } else {
+                        Mutability::Mut
```

Comment:
We could have an extra check for `matches!(self.ctfe_mode, Some(Const(..)))` which then rejects mutable references immediately... but we already intern all allocations as immutable and so such code will already error, just with a different error message ("mutable ref in const" vs "dangling ref"/"mutable ref to immutable memory").